fix commands crash, panic: send on closed channel

Thumbs up 👍  this issue if you've seen a panic like this. And contact me!

also <@178341568279871500> 
found a panic triggering a command

```

[ERROR] 2022/07/26 08:46:10 http: panic serving 127.0.0.1:56134: send on closed channel
goroutine 239 [running]:
net/http.(*conn).serve.func1()
    net/http/server.go:1825 +0xbf
panic({0xc3a2a0, 0xe3da50})
    runtime/panic.go:844 +0x258

```

logs and dets
it was pmm update triggered
<https://gist.github.com/bakerboy448/cbb877692714aa73762ad522a50cf6ac>

Origin user: bakerboy448
Origin message: [Discord Server Link](https://discord.com/channels/764440599066574859/843676397099679784/1001490337371529247)
Reporting user: TwitchCaptain
Reporting message: [Discord Server Link](https://discord.com/channels/764440599066574859/843676397099679784/1001542484603715724)

--- a/init/docker/Dockerfile
+++ b/init/docker/Dockerfile
@@ -12,14 +12,9 @@ COPY go.mod go.sum .
 RUN go mod download
 COPY . .
 
-ARG BINARY=application-builder
-ARG BUILD_DATE=0
-ARG COMMIT=0
-ARG VERSION=unknown
 ARG TARGETOS
 ARG BINARY
 ARG TARGETARCH
-ARG BUILD_FLAGS
 
 # For megacli. All the *'s are to deal with multiarch. :(
 RUN apt update && apt install -y libncurses5 libstdc++6 libtinfo5 apt-utils && \

--- a/init/docker/build
+++ b/init/docker/build
@@ -5,6 +5,7 @@ pushd ../..
 
 source settings.sh
 
+docker buildx create --use
 
 if [ "v$VERSION" = "$SOURCE_BRANCH" ]; then
   TAGS="--tag ${DOCKER_REPO}:$VERSION"

--- a/pkg/bindata/templates/landing.html
+++ b/pkg/bindata/templates/landing.html
@@ -11,7 +11,7 @@
                                                 If you have more ideas for this interface, or find any bugs,
                                                 <a href="https://github.com/Notifiarr/notifiarr/issues/new">please let us know!</a>
                                             </p>
-                                            <p><b>{{.ClientInfo.String}}</b></p>
+                                            <p>{{todaysemoji}} <b>{{.ClientInfo.String}}</b></p>
                                             <p class="text-center"><img src="{{files}}/images/golift.png" style="height:150px;"></p>
                                         </div>
                                         <div class="col-md-1">

--- a/pkg/bindata/templates/system.html
+++ b/pkg/bindata/templates/system.html
@@ -67,7 +67,7 @@
                                              </tr>
                                              <tr>
                                                  <td>Version</td>
-                                                 <td>{{.Version.version}}.{{.Version.revision}}</td>
+                                                 <td>{{.Version.version}}-{{.Version.revision}}</td>
                                                  <td style="display:none;"></td>
                                              </tr>
                                              <tr>

--- a/pkg/client/html_templates.go
+++ b/pkg/client/html_templates.go
@@ -111,7 +111,8 @@ func (c *Client) getFuncMap() template.FuncMap {
 
 			return date.String()
 		},
-		"fortune": Fortune,
+		"todaysemoji": mnd.TodaysEmoji,
+		"fortune":     Fortune,
 		// returns the current time.
 		"now": time.Now,
 		// returns an integer divided by a million.

--- a/pkg/client/html_templates.go
+++ b/pkg/client/html_templates.go
@@ -143,7 +143,7 @@ func (c *Client) getFuncMap() template.FuncMap {
 			}
 			return "0"
 		},
-		"one259": func() (num []float64) {
+		"one259": func() (num []float64) { // 1 to 59
 			for i := float64(1); i < 60; i++ {
 				num = append(num, i)
 			}

--- a/pkg/client/start.go
+++ b/pkg/client/start.go
@@ -7,7 +7,6 @@ package client
 
 import (
 	"context"
-	"errors"
 	"fmt"
 	"html/template"
 	"net/http"
@@ -24,6 +23,7 @@ import (
 	"github.com/Notifiarr/notifiarr/pkg/update"
 	"github.com/Notifiarr/notifiarr/pkg/website"
 	"github.com/gorilla/securecookie"
+	"github.com/hako/durafmt"
 	flag "github.com/spf13/pflag"
 	"golift.io/version"
 )
@@ -120,8 +120,9 @@ func (c *Client) start() error { //nolint:cyclop
 	}
 
 	c.Logger.SetupLogging(c.Config.LogConfig)
-	c.Printf("%s v%s-%s Starting! [PID: %v] %v",
-		c.Flags.Name(), version.Version, version.Revision, os.Getpid(), time.Now())
+	c.Printf(" %s %s v%s-%s Starting! [PID: %v] %s",
+		mnd.TodaysEmoji(), c.Flags.Name(), version.Version, version.Revision, os.Getpid(),
+		version.Started.Format("Monday, January 2, 2006 @ 3:04:05 PM MST -0700"))
 	c.Printf("==> %s", msg)
 
 	c.printUpdateMessage()
@@ -245,6 +246,12 @@ func (c *Client) triggerConfigReload(event website.EventType, source string) {
 
 // Exit stops the web server and logs our exit messages. Start() calls this.
 func (c *Client) Exit() error {
+	defer func() {
+		defer c.CapturePanic()
+		//nolint:gomnd
+		c.Print(" ❌ Good bye! Uptime:", durafmt.Parse(time.Since(version.Started).Round(time.Second)).LimitFirstN(3))
+	}()
+
 	c.StartWebServer()
 	c.setSignals()
 
@@ -257,7 +264,7 @@ func (c *Client) Exit() error {
 			}
 		case sigc := <-c.sigkil:
 			c.Printf("[%s] Need help? %s\n=====> Exiting! Caught Signal: %v", c.Flags.Name(), mnd.HelpLink, sigc)
-			return c.exit()
+			return c.stop(website.EventSignal)
 		case sigc := <-c.sighup:
 			if err := c.checkReloadSignal(sigc); err != nil {
 				return err // reloadConfiguration()
@@ -266,40 +273,18 @@ func (c *Client) Exit() error {
 	}
 }
 
-// This is called from at least two different exit points.
-func (c *Client) exit() error {
-	if c.server == nil {
-		return nil
-	}
-
-	ctx, cancel := context.WithTimeout(context.Background(), c.Config.Timeout.Duration)
-	defer cancel()
-
-	if err := c.server.Shutdown(ctx); err != nil {
-		return fmt.Errorf("server shutdown failed: %w", err)
-	}
-
-	c.server = nil
-
-	return nil
-}
-
 // reloadConfiguration is called from a menu tray item or when a HUP signal is received.
 // Re-reads the configuration file and stops/starts all the internal routines.
 // Also closes and re-opens all log files. Any errors cause the application to exit.
 func (c *Client) reloadConfiguration(event website.EventType, source string) error {
 	c.Printf("==> Reloading Configuration (%s): %s", event, source)
-	c.closeDynamicTimerMenus()
-	c.triggers.Stop(event)
-	c.Config.Services.Stop()
-	c.website.Stop()
 
-	err := c.StopWebServer()
-	if err != nil && !errors.Is(err, ErrNoServer) {
+	err := c.stop(event)
+	if err != nil {
 		return fmt.Errorf("stoping web server: %w", err)
-	} else if err == nil {
-		defer c.StartWebServer()
 	}
+
+	defer c.StartWebServer()
 
 	// start over.
 	c.Config = configfile.NewConfig(c.Logger)
@@ -314,11 +299,25 @@ func (c *Client) reloadConfiguration(event website.EventType, source string) err
 	c.Logger.SetupLogging(c.Config.LogConfig)
 	clientInfo := c.configureServices()
 	c.setupMenus(clientInfo)
-	c.Print("==> Configuration Reloaded! Config File:", c.Flags.ConfigFile)
+	c.Print(" 🔄 Configuration Reloaded! Config File:", c.Flags.ConfigFile)
 
 	if err = ui.Notify("Configuration Reloaded! Config File: %s", c.Flags.ConfigFile); err != nil {
 		c.Errorf("Creating Toast Notification: %v", err)
 	}
 
 	return nil
+}
+
+// stop is called from at least two different exit points and on reload.
+func (c *Client) stop(event website.EventType) error {
+	defer func() {
+		defer c.CapturePanic()
+		c.closeDynamicTimerMenus()
+		c.triggers.Stop(event)
+		c.Config.Services.Stop()
+		c.website.Stop()
+		c.Print("==> All systems powered down!")
+	}()
+
+	return c.StopWebServer()
 }

--- a/pkg/client/tray.go
+++ b/pkg/client/tray.go
@@ -48,7 +48,7 @@ func (c *Client) startTray(clientInfo *website.ClientInfo) {
 		}
 	}, func() {
 		// This code only fires from menu->quit.
-		if err := c.exit(); err != nil {
+		if err := c.stop(website.EventUser); err != nil {
 			c.Errorf("Server: %v", err)
 			os.Exit(1) // web server problem
 		}

--- a/pkg/client/webserver.go
+++ b/pkg/client/webserver.go
@@ -16,9 +16,6 @@ import (
 	apachelog "github.com/lestrrat-go/apache-logformat"
 )
 
-// ErrNoServer returns when the server is already stopped and a stop req occurs.
-var ErrNoServer = fmt.Errorf("the web server is not running, cannot stop it")
-
 // StartWebServer starts the web server.
 func (c *Client) StartWebServer() {
 	c.Lock()
@@ -81,6 +78,8 @@ func (c *Client) runWebServer() {
 
 // StopWebServer stops the web servers. Panics if that causes an error or timeout.
 func (c *Client) StopWebServer() error {
+	c.Print("==> Stopping Web Server!")
+
 	ctx, cancel := context.WithTimeout(context.Background(), c.Config.Timeout.Duration)
 	defer cancel()
 
@@ -125,7 +124,7 @@ func (r *responseWrapper) Write(b []byte) (int, error) {
 func (r *responseWrapper) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	hijack, ok := r.ResponseWriter.(http.Hijacker)
 	if !ok {
-		return nil, nil, ErrNoServer
+		panic("cannot hijack connection!")
 	}
 
 	conn, buf, err := hijack.Hijack()

--- a/pkg/mnd/constants.go
+++ b/pkg/mnd/constants.go
@@ -21,7 +21,7 @@ const (
 	Bits32    = 32
 	Windows   = "windows"
 	Disabled  = "disabled"
-	HelpLink  = "GoLift Discord: https://golift.io/discord"
+	HelpLink  = "Notifiarr Discord: https://notifiarr.com/discord"
 	UserRepo  = "Notifiarr/notifiarr"
 	BugIssue  = "This is a bug please report it on github: https://github.com/" + UserRepo + "/issues/new"
 	DockerV   = "NOTIFIARR_IN_DOCKER"

--- a/pkg/mnd/emoji.go
+++ b/pkg/mnd/emoji.go
@@ -15,12 +15,13 @@ const (
 func TodaysEmoji() string {
 	today := version.Started.YearDay()
 
-	if leapYear(version.Started.Year()) {
-		if today == leapDay {
-			today = altLeapDay
-		} else {
-			today--
-		}
+	switch year := version.Started.Year(); {
+	case !leapYear(year), today < leapDay:
+		break
+	case today == leapDay:
+		today = altLeapDay
+	default:
+		today--
 	}
 
 	if emoji, ok := specialDays[today]; ok {
@@ -59,5 +60,4 @@ var specialDays = map[int]string{ //nolint:gochecknoglobals
 	328:        "🦃", // November 24
 	359:        "🎄", // December 25
 	altLeapDay: "🤹", // February 29 (Leap Day)
-
 }

--- a/pkg/mnd/emoji.go
+++ b/pkg/mnd/emoji.go
@@ -7,57 +7,58 @@ import (
 )
 
 const (
-	leapDay    = 60
+	leapDay    = 60 // day of year leap day falls on.
 	altLeapDay = 366
 )
 
-// TodaysEmoji returns an emoji specific to the month (or perhaps date).
-func TodaysEmoji() string {
-	today := version.Started.YearDay()
-
-	switch year := version.Started.Year(); {
-	case !leapYear(year), today < leapDay:
-		break
+func today(when time.Time) int {
+	switch today := when.YearDay(); {
+	case !leapYear(when.Year()), today < leapDay:
+		return today
 	case today == leapDay:
-		today = altLeapDay
+		return altLeapDay
 	default:
-		today--
+		return today - 1
 	}
-
-	if emoji, ok := specialDays[today]; ok {
-		return emoji
-	}
-
-	return monthEmojis[version.Started.Month()]
 }
 
 func leapYear(year int) bool {
 	return year%400 == 0 || (year%4 == 0 && year%100 != 0)
 }
 
-var monthEmojis = map[time.Month]string{ //nolint:gochecknoglobals
-	time.January:   "🤖", //
-	time.February:  "😻", //
-	time.March:     "🗼", //
-	time.April:     "🌦", //
-	time.May:       "🌸", //
-	time.June:      "🍀", //
-	time.July:      "🌵", //
-	time.August:    "🔥", //
-	time.September: "🍁", //
-	time.October:   "🍉", //
-	time.November:  "🍗", //
-	time.December:  "⛄", //
+func emojiMonth(when time.Time) string {
+	return map[time.Month]string{
+		time.January:   "🤖", //
+		time.February:  "😻", //
+		time.March:     "🗼", //
+		time.April:     "🌧", //
+		time.May:       "🌸", //
+		time.June:      "🍄", //
+		time.July:      "🌵", //
+		time.August:    "🔥", //
+		time.September: "🐸", //
+		time.October:   "🍁", //
+		time.November:  "👽", //
+		time.December:  "⛄", //
+	}[when.Month()]
 }
 
-var specialDays = map[int]string{ //nolint:gochecknoglobals
-	1:          "🎉", // January 1
-	45:         "💝", // February 14
-	185:        "🧨", // July 4
-	229:        "🏄", // August 17
-	304:        "🎃", // October 31
-	315:        "🪖", // November 11
-	328:        "🦃", // November 24
-	359:        "🎄", // December 25
-	altLeapDay: "🤹", // February 29 (Leap Day)
+// TodaysEmoji returns an emoji specific to the month (or perhaps date).
+func TodaysEmoji() string {
+	if emoji, ok := map[int]string{
+		1:          "🎉", // January 1
+		45:         "💝", // February 14
+		185:        "🧨", // July 4
+		229:        "🏄", // August 17
+		254:        "⛑", // September 11
+		304:        "🎃", // October 31
+		315:        "🪖", // November 11
+		328:        "🦃", // November 24
+		359:        "🎄", // December 25
+		altLeapDay: "🤹", // February 29 (Leap Day)
+	}[today(version.Started)]; ok {
+		return emoji
+	}
+
+	return emojiMonth(version.Started)
 }

--- a/pkg/mnd/emoji.go
+++ b/pkg/mnd/emoji.go
@@ -1,0 +1,63 @@
+package mnd
+
+import (
+	"time"
+
+	"golift.io/version"
+)
+
+const (
+	leapDay    = 60
+	altLeapDay = 366
+)
+
+// TodaysEmoji returns an emoji specific to the month (or perhaps date).
+func TodaysEmoji() string {
+	today := version.Started.YearDay()
+
+	if leapYear(version.Started.Year()) {
+		if today == leapDay {
+			today = altLeapDay
+		} else {
+			today--
+		}
+	}
+
+	if emoji, ok := specialDays[today]; ok {
+		return emoji
+	}
+
+	return monthEmojis[version.Started.Month()]
+}
+
+func leapYear(year int) bool {
+	return year%400 == 0 || (year%4 == 0 && year%100 != 0)
+}
+
+var monthEmojis = map[time.Month]string{ //nolint:gochecknoglobals
+	time.January:   "🤖", //
+	time.February:  "😻", //
+	time.March:     "🗼", //
+	time.April:     "🌦", //
+	time.May:       "🌸", //
+	time.June:      "🍀", //
+	time.July:      "🌵", //
+	time.August:    "🔥", //
+	time.September: "🍁", //
+	time.October:   "🍉", //
+	time.November:  "🍗", //
+	time.December:  "⛄", //
+}
+
+var specialDays = map[int]string{ //nolint:gochecknoglobals
+	1:          "🎉", // January 1
+	45:         "💝", // February 14
+	185:        "🧨", // July 4
+	229:        "🏄", // August 17
+	304:        "🎃", // October 31
+	315:        "🪖", // November 11
+	328:        "🦃", // November 24
+	359:        "🎄", // December 25
+	altLeapDay: "🤹", // February 29 (Leap Day)
+
+}

--- a/pkg/services/apps.go
+++ b/pkg/services/apps.go
@@ -1,6 +1,7 @@
 package services
 
 import (
+	"net/url"
 	"strings"
 
 	"golift.io/cnfg"
@@ -194,9 +195,10 @@ func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funl
 
 		prefix := "" // add auth to the url here. woo, hacky, but it works!
 		if !strings.Contains(app.Config.URL, "@") {
-			prefix = "http://" + app.User + ":" + app.Pass + "@"
+			user := url.PathEscape(app.User) + ":" + url.PathEscape(app.Pass) + "@"
+			prefix = "http://" + user
 			if strings.HasPrefix(app.Config.URL, "https://") {
-				prefix = "https://" + app.User + ":" + app.Pass + "@"
+				prefix = "https://" + user
 			}
 		}
 

--- a/pkg/services/apps.go
+++ b/pkg/services/apps.go
@@ -194,10 +194,10 @@ func (c *Config) collectDownloadApps(svcs []*Service) []*Service { //nolint:funl
 		}
 
 		prefix := "" // add auth to the url here. woo, hacky, but it works!
+
 		if !strings.Contains(app.Config.URL, "@") {
 			user := url.PathEscape(app.User) + ":" + url.PathEscape(app.Pass) + "@"
-			prefix = "http://" + user
-			if strings.HasPrefix(app.Config.URL, "https://") {
+			if prefix = "http://" + user; strings.HasPrefix(app.Config.URL, "https://") {
 				prefix = "https://" + user
 			}
 		}

--- a/pkg/services/config.go
+++ b/pkg/services/config.go
@@ -49,6 +49,7 @@ type Config struct {
 	triggerChan  chan website.EventType
 	checkChan    chan triggerCheck
 	stopLock     sync.Mutex
+	lastUpdate   time.Time // last time we sent service states to website.
 }
 
 // CheckType locks us into a few specific types of checks.

--- a/pkg/services/interface.go
+++ b/pkg/services/interface.go
@@ -55,9 +55,9 @@ func (c *Config) runCheck(svc *Service, force bool) bool {
 }
 
 // runChecks runs checks that are due. Passing true, runs them even if they're not due.
-func (c *Config) runChecks(forceAll bool) {
+func (c *Config) runChecks(forceAll bool) bool {
 	if c.checks == nil || c.done == nil {
-		return
+		return false
 	}
 
 	count := 0
@@ -79,12 +79,14 @@ func (c *Config) runChecks(forceAll bool) {
 		somethingChanged = <-c.done || somethingChanged
 	}
 
-	if somethingChanged {
-		c.updateStatesOnSite()
-	}
+	return somethingChanged
 }
 
-func (c *Config) updateStatesOnSite() {
+func (c *Config) updateStatesOnSite(force bool) {
+	if !force && time.Since(c.lastUpdate) < time.Hour {
+		return
+	}
+
 	var (
 		err    error
 		values = make(map[string][]byte)
@@ -102,11 +104,15 @@ func (c *Config) updateStatesOnSite() {
 	}
 
 	if len(values) == 0 {
+		c.Printf("No service states to send updates to website? Weird.")
 		return
 	}
 
 	if err := c.Website.SetValues(values); err != nil {
-		c.Errorf("Setting Service States on Notifiarr.com: %v", err)
+		c.Errorf("Setting Service States on website: %v", err)
+	} else {
+		c.lastUpdate = time.Now()
+		c.Printf("==> Updated [internal] Service States on website.")
 	}
 }
 

--- a/pkg/services/services.go
+++ b/pkg/services/services.go
@@ -128,8 +128,10 @@ func (c *Config) loadServiceStates() {
 					break
 				}
 
-				if time.Since(svc.LastCheck) < time.Hour {
-					c.Debugf("Updated initial service state with website-saved data: %s, %s", name, svc.State)
+				if time.Since(svc.LastCheck) < 2*time.Hour {
+					c.Printf("==> Set service state with website-saved data: %s, %s for %s",
+						name, svc.State, time.Since(svc.Since).Round(time.Second))
+
 					c.services[name].svc.Output = svc.Output
 					c.services[name].svc.State = svc.State
 					c.services[name].svc.Since = svc.Since
@@ -177,10 +179,10 @@ func (c *Config) runServiceChecker() { //nolint:cyclop
 		case event := <-c.checkChan:
 			c.Debugf("Running service check '%s' via event: %s, buffer: %d/%d",
 				event.Service.Name, event.Source, len(c.checks), cap(c.checks))
-			c.runCheck(event.Service, true)
+			c.updateStatesOnSite(c.runCheck(event.Service, true))
 		case event := <-c.triggerChan:
 			c.Debugf("Running all service checks via event: %s, buffer: %d/%d", event, len(c.checks), cap(c.checks))
-			c.runChecks(true)
+			c.updateStatesOnSite(c.runChecks(true))
 
 			if event != "log" {
 				c.SendResults(&Results{What: event, Svcs: c.GetResults()})
@@ -195,7 +197,7 @@ func (c *Config) runServiceChecker() { //nolint:cyclop
 
 			c.Debug("Service Checks Payload (log only):", string(data))
 		case <-second.C:
-			c.runChecks(false)
+			c.updateStatesOnSite(c.runChecks(false))
 		}
 	}
 }
@@ -212,7 +214,7 @@ func (c *Config) Stop() {
 	c.stopLock.Lock()
 	defer c.stopLock.Unlock()
 
-	c.updateStatesOnSite()
+	c.updateStatesOnSite(true)
 
 	if c.stopChan == nil {
 		return

--- a/pkg/website/website.go
+++ b/pkg/website/website.go
@@ -192,6 +192,11 @@ func readBodyForLog(body io.Reader, max int64) string {
 }
 
 func (s *Server) watchSendDataChan() {
+	defer func() {
+		defer s.config.CapturePanic()
+		s.config.Printf("==> Website notifier shutting down. No more ->website requests may be sent!")
+	}()
+
 	for data := range s.sendData {
 		switch resp, elapsed, err := s.sendRequest(data); {
 		case data.LogMsg == "":

